## Document Objects

This Part of the lab is about using "Document Objects". <br>
Basically, Document objects are just wrappers/containers for large texts (not necessarily large, but yeah, we use it for very large texts, such as those coming from Docs, or pdfs).<br>
And its necessary, as raw strings just store text as, you know, just "text"..<br>
But when the Document is too large, and texts are too much, then, we need a wrapper, which not only takes the text, but also stores them in an organized and labelled manner.<br>
Like, this is the format for a Document Object:<br>
document_object= =  Document(<br>
    page_content="Python is a programming language...",<br>
    metadata={"source": "python.pdf", "page": 1}<br>
)<br>
.......<br>
As we can see here, it has the 'page_content' and also the 'metadata'. <br>
Using just Plain Strings will just store the big text, but it'll be very hard to understand from where a particular text came from.<br>
On the other hand, since 'Document objects' store texts in an organized way, it divides the huge texts from a document in 'chunks'. So even if we have 100 chunks from a PDF, we can easily know:<br>
Which chunk came from which document?<br>
Which page was it on?<br>
What was the source URL?<br>
Without metadata we'd just have 100 anonymous strings with no way to trace them back.<br>
So when the AI answers a question using chunk 47 from our PDF, it'll be able to say: "This answer came from page 3 of someuploaded-paper.pdf"<br>
That's only possible because the chunk is wrapped in a Document object with metadata tracking where it came from.<br>
#So Document objects serve two purposes: They Sarry the text through the pipeline & Track where it came from via metadata.<br>
..........................................<br>


In [3]:
%%capture
!pip install --force-reinstall --no-cache-dir tenacity==8.2.3 --user
!pip install "ibm-watsonx-ai==1.0.8" --user
!pip install "ibm-watson-machine-learning==1.0.367" --user
!pip install "langchain-ibm==0.1.7" --user
!pip install "langchain-community==0.2.10" --user
!pip install "langchain-experimental==0.0.62" --user
!pip install "langchainhub==0.1.18" --user
!pip install "langchain==0.2.11" --user
!pip install "pypdf==4.2.0" --user
!pip install "chromadb==0.4.24" --user

In [ ]:
import os
os._exit(00)

In [1]:
# You can also use this section to suppress warnings generated by your code:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn
warnings.filterwarnings('ignore')
import os
os.environ['ANONYMIZED_TELEMETRY'] = 'False'

from ibm_watsonx_ai.foundation_models import ModelInference
from ibm_watsonx_ai.metanames import GenTextParamsMetaNames as GenParams
from ibm_watsonx_ai.foundation_models.utils.enums import ModelTypes
from ibm_watson_machine_learning.foundation_models.extensions.langchain import WatsonxLLM

Okay, So now, in the section below, have made our document object named 'doc1', which stores a particular content and its metadata.

In [3]:
# Import the Document class from langchain_core.documents module
# Document is a container for text content with associated metadata
from langchain_core.documents import Document

# Create a Document instance with:
# 1. page_content: The actual text content about Python
# 2. metadata: A dictionary containing additional information about this document
doc1 = Document(page_content="""Python is an interpreted high-level general-purpose programming language.
 Python's design philosophy emphasizes code readability with its notable use of significant indentation.""",
metadata={
    'my_document_id' : 234234,                      # Unique identifier for this document
    'my_document_source' : "About Python",          # Source or title information
    'my_document_create_time' : 1680013019          # Unix timestamp for document creation (March 28, 2023)
 })

The data stored in a particular document object (here doc1) can be printed in various ways, depending on how we want it.<br>
Fo example:
1. To just print the content: print(document object.page_content)<br>
2. To just print the metadata: print(document object.metadata)<br>
3. To Print the page content, but as raw with just continous lines and literally \n(line breakers): document object.page_content (We just called it)<br>
4. Simply calling the document object will lead to printing everything it contains in raw format.<br>

In [4]:
print(doc1.page_content)

Python is an interpreted high-level general-purpose programming language.
 Python's design philosophy emphasizes code readability with its notable use of significant indentation.


In [5]:
print(doc1.metadata)

{'my_document_id': 234234, 'my_document_source': 'About Python', 'my_document_create_time': 1680013019}


In [6]:
doc1.page_content

"Python is an interpreted high-level general-purpose programming language.\n Python's design philosophy emphasizes code readability with its notable use of significant indentation."

In [7]:
doc1

Document(metadata={'my_document_id': 234234, 'my_document_source': 'About Python', 'my_document_create_time': 1680013019}, page_content="Python is an interpreted high-level general-purpose programming language.\n Python's design philosophy emphasizes code readability with its notable use of significant indentation.")

## Document Loaders: PDF Loader and URL Loader
What is a Document loader?<br>
Basically, if the doc is in the form of PDF in the internet, we can store it in a document loader using PyPDFLoader. <br>
Similarly, if the doc is in the form of a url or website, we can store it by using WebBaseLoader.<br>
Then, this loader is loaded/saved into the Document object, in the format: <br>
document object = document loader.load() <br>
Now, we can call or print the content in that document accordingly.<br>
For Example;<br>
1. Printing content from page 2 of a document, but in raw form: document[2]<br>
2. Printing the first 1000 words from page 1 of the document loaded, but raw: document[1].page_content[:1000]<br>
3. Printing the first 1000 words from page 1 of the document loaded, but in the proper format, just like in the document: print(document[1].page_content[:1000])<br>

## PDF Loader

In [9]:
# Import the PyPDFLoader class from langchain_community's document_loaders module
# This loader is specifically designed to load and parse PDF files
from langchain_community.document_loaders import PyPDFLoader

# We create a PyPDFLoader instance by passing the URL of the PDF file
# The loader will download the PDF from the specified URL and prepare it for loading
loader = PyPDFLoader("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf")

# Call the load() method to:
# 1. Download the PDF if needed
# 2. Extract text from each page
# 3. Create a list of Document objects, one for each page of the PDF
# Each Document will contain the text content of a page and metadata including page number
document = loader.load()

In [10]:
document[2]  # take a look at the page 2

Document(metadata={'source': 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf', 'page': 2}, page_content=' \nFigure 2. An AIMessage illustration  \nC. Prompt Template  \nPrompt templates  [10] allow you to structure  input for LLMs. \nThey provide a convenient way to format user inputs and \nprovide instructions to generate responses. Prompt templates \nhelp ensure that the LLM understands the  desired context and \nproduces relevant outputs.  \nThe prompt template classes in LangChain  are built to \nmake constructing prompts with dynamic inputs easier. Of \nthese classes, the simplest is the PromptTemplate.  \nD. Chain  \nChains  [11] in LangChain refer to the combination of \nmultiple components to achieve specific tasks. They provide \na structured and modular approach to building language \nmodel applications. By combining different components, you \ncan create chains that address various u se cases and \nrequirements. 

In [11]:
print(document[1].page_content[:1000])  # print the page 1's first 1000 tokens

LangChain helps us to unlock the ability to harness the 
LLM’s immense potential in tasks such as document analysis, 
chatbot development, code analysis, and countless other 
applications. Whether your desire is to unlock deeper natural 
language understanding , enhance data, or circumvent 
language barriers through translation, LangChain is ready to 
provide the tools and programming support you need to do 
without it that it is not only difficult but also fresh for you . Its 
core functionalities encompass:  
1. Context -Aware Capabilities: LangChain facilitates the 
development of applications that are inherently 
context -aware. This means that these applications can 
connect to a language model and draw from various 
sources of context, such as prompt instructions, a  few-
shot examples, or existing content, to ground their 
responses effectively.  
2. Reasoning Abilities: LangChain equips applications 
with the capacity to reason effectively. By relying on a 
language model, thes

In [12]:
document[1].page_content[:1000]

'LangChain helps us to unlock the ability to harness the \nLLM’s immense potential in tasks such as document analysis, \nchatbot development, code analysis, and countless other \napplications. Whether your desire is to unlock deeper natural \nlanguage understanding , enhance data, or circumvent \nlanguage barriers through translation, LangChain is ready to \nprovide the tools and programming support you need to do \nwithout it that it is not only difficult but also fresh for you . Its \ncore functionalities encompass:  \n1. Context -Aware Capabilities: LangChain facilitates the \ndevelopment of applications that are inherently \ncontext -aware. This means that these applications can \nconnect to a language model and draw from various \nsources of context, such as prompt instructions, a  few-\nshot examples, or existing content, to ground their \nresponses effectively.  \n2. Reasoning Abilities: LangChain equips applications \nwith the capacity to reason effectively. By relying on a \nl

## URL Loader

In [15]:
# Import the WebBaseLoader class from langchain_community's document_loaders module
# This loader is designed to scrape and extract text content from web pages
from langchain_community.document_loaders import WebBaseLoader

# Create a WebBaseLoader instance by passing the URL of the web page to load
# This URL points to the LangChain documentation's introduction page
loader = WebBaseLoader("https://python.langchain.com/v0.2/docs/introduction/")

# Call the load() method to:
# 1. Send an HTTP request to the specified URL
# 2. Download the HTML content
# 3. Parse the HTML to extract meaningful text
# 4. Create a list of Document objects containing the extracted content
web_data = loader.load()

# Print the first 1000 characters of the page content from the first Document
# This provides a preview of the successfully loaded web content
# web_data[0] accesses the first Document in the list
# .page_content accesses the text content of that Document
# [:1000] slices the string to get only the first 1000 characters
print(web_data[0].page_content[:1000])

LangChain overview - Docs by LangChainDocumentation IndexFetch the complete documentation index at: /llms.txtUse this file to discover all available pages before exploring further.Skip to main contentJoin our product experts for a "Build More with LangSmith: Summer AMA Series" from July 9-August 12, 2026. RSVP todayDocs by LangChain home pageBuildSearch...⌘KAsk AIGitHubTry LangSmithTry LangSmithSearch...NavigationLangChain overviewOverviewDeep AgentsLangChainLangGraphIntegrationsLearnReferenceContributePythonOverviewGet startedInstallQuickstartChangelogPhilosophyCore componentsAgentsModelsMessagesToolsShort-term memoryEvent streamingStreamingStructured outputMiddlewareOverviewPrebuilt middlewareCustom middlewareFrontendOverviewPatternsIntegrationsAdvanced usageGuardrailsRuntimeContext engineeringModel Context Protocol (MCP)Human-in-the-loopMulti-agentRetrievalLong-term memoryAgent developmentLangSmith StudioTestAgent Chat UIProductionDeploymentObservabilityOn this page Create an agent 

................................................................................ <br>
So till here, we have understood Document Objects, Document loaders, and how to print particular text from a vast document. <br>
But now, the question is: <br>
Can we decide the size of the chunks/the way a document is being divided? <br>
Like till now, chunks our being made automatically. The PDF already has natural page boundaries so the loader uses those. <br>
Page 1 → Document[0] <br>
Page 2 → Document[1] <br>
Page 3 → Document[2] <br>
No intelligence here — just follows page breaks. <br>
................ <br>
So, to divide the vast doc according to us, we may use "Text Splitters". <br>
Format is given as: <br>
text_splitter = CharacterTextSplitter(chunk_size=200, chunk_overlap=20) <br>
chunks = text_splitter.split_documents(document) <br>
This splits by character count — not by topics or chapters. It doesn't understand the content at all. It just counts characters and cuts at \n whenever it gets close to 200 characters. <br>
...<br>
How Do Text Splitters Work?<br>
Text Splitters are crucial for RAG. A PDF might be 50 pages. You can't send all 50 pages to the AI at once because:<br>
1. LLMs have a token limit (context window)<br>
2. Sending everything is slow and expensive<br>
3. The AI gets confused with too much information<br>

So we split it into small chunks:<br>
Example:<br>
Page 1 (2000 words)<br>
    ↓ split<br>
Chunk 1 (200 chars)<br>
Chunk 2 (200 chars)  ← 20 char overlap with chunk 1<br>
Chunk 3 (200 chars)  ← 20 char overlap with chunk 2<br>
The overlap is important as it makes sure context isn't lost at the boundary between chunks.<br>
It happens something like:<br>
Chunk 1: "...LangChain is a framework for building"<br>
Chunk 2: "for building AI applications using large language models..."<br>
.......<br>
There are 3 types of Text Splitters: <br>
1. CharacterTextSplitter → splits by character count (best for: Clean, well-formatted text with consistent newlines) <br>
2. RecursiveCharacterTextSplitter → tries paragraphs first, then sentences, then characters (best for: Most real-world use cases — messy web pages, mixed formatting. Most commonly used in RAG) <br>
3. TokenTextSplitter → splits by token count (best for: When you need precise token counts — important when working close to model context limits) <br>

## Text Splitter

## 1. CharacterTextSplitter
CharacterTextSplitter is the simplest method of splitting the content. These splits are based on characters (by default "\n\n") and measures chunk length by number of characters.

In [22]:
# Import the CharacterTextSplitter class from langchain.text_splitter module
# Text splitters are used to divide large texts into smaller, manageable chunks
from langchain.text_splitter import CharacterTextSplitter

# Create a CharacterTextSplitter with specific configuration:
# - chunk_size=200: Each chunk will contain approximately 200 characters
# - chunk_overlap=20: Consecutive chunks will overlap by 20 characters to maintain context
# - separator="\n": Text will be split at newline characters when possible
text_splitter = CharacterTextSplitter(chunk_size=200, chunk_overlap=20, separator="\n")

# Split the previously loaded document (PDF or other text) into chunks
# The split_documents method:
# 1. Takes a list of Document objects
# 2. Splits each document's content based on the configured parameters
# 3. Returns a new list of Document objects where each contains a chunk of text
# 4. Preserves the original metadata for each chunk
chunk1 = text_splitter.split_documents(document)
# Print the total number of chunks created
# This shows how many smaller Document objects were generated from the original document(s)
# The number depends on the original document length and the chunk_size setting
print(len(chunk1))

148


In [19]:
chunk1[5].page_content   # take a look at any chunk's page content

'contextualized language models to introduce MindGuide, an \ninnovative chatbot serving as a mental health assistant for \nindividuals seeking guidance and support in these critical areas.'

In [27]:
# Just to check how the text splitter works
print(document[0].page_content[:1000])

* corresponding author - jkim72@kent.edu  Revolutionizing Mental Health Care through 
LangChain: A Journey with a Large Language 
Model
Aditi Singh  
 Computer Science   
 Cleveland State University   
 a.singh22@csuohio.edu  Abul Ehtesham   
The Davey Tree Expert 
Company   
abul.ehtesham@davey.com  Saifuddin Mahmud  
Computer Science & 
Information Systems   
 Bradley University  
smahmud@bradley.edu   Jong -Hoon Kim * 
 Computer Science,  
Kent State University,  
jkim72@kent.edu  
Abstract— Mental health challenges are on the rise in our 
modern society, and the imperative to address mental disorders, 
especially regarding anxiety, depression, and suicidal thoughts, 
underscores the need for effective interventions. This paper 
delves into the ap plication of recent advancements in pretrained 
contextualized language models to introduce MindGuide, an 
innovative chatbot serving as a mental health assistant for 
individuals seeking guidance and support in these critical areas. 
Mind

## NOTES:
I could understand how the 3 types of text splitters work, so here is what I got from the internet:<br>
Text splitters break massive documents into smaller, manageable chunks so Large Language Models (LLMs) can process them without exceeding their context limits and generate higher-quality outputs.<br>
Here is how each of the three splitters processes text: <br>
1. CharacterTextSplitter → Splits by character count<br>
This is the simplest and most rigid method. It divides text mechanically based on a hardcoded number of characters (like a string length in programming).<br>
How it works: If you set your chunk size to (100) characters, it counts exactly (100) characters and slices the text, regardless of whether it is in the middle of a sentence or a word.<br>
Best used for: Highly structured, plain text where you need absolute, consistent length control.<br>
2. RecursiveCharacterTextSplitter → Tries paragraphs, sentences, then characters<br>
This is the default and recommended choice for most use cases. Instead of bluntly cutting off text, it looks at the natural structure of your document.<br>
How it works: It tries to split using large boundaries first (like double newlines, which indicate paragraphs). If the resulting paragraph chunk is still too big, it moves to the next smaller boundary (single newlines or sentences), then words, and finally individual characters.<br>
Best used for: Standard documents, books, and articles. It keeps whole thoughts, paragraphs, and sentences intact as much as possible.<br>
3. TokenTextSplitter → Splits by token count<br>
LLMs don't read words or characters; they read "tokens" (chunks of words or sub-words, where (1) token  is roughly 4 characters or (0.75) words). This method splits text exactly by how LLMs understand it.<br>
How it works: It breaks text according to an LLM tokenizer. If your limit is (1000) tokens, it ensures each chunk is strictly under (1000) tokens, preventing truncation errors in the AI model.<br>
Best used for: Environments where you are dealing with strict model input/context limits and need flawless compatibility with tools like OpenAI's models or Claude.<br>

In [31]:
# Now let us try this:
chunk2 = text_splitter.split_documents(web_data)
print(len(chunk2))

Created a chunk of size 1544, which is longer than the specified 200
Created a chunk of size 730, which is longer than the specified 200
Created a chunk of size 835, which is longer than the specified 200


10


...................... <br>
Why chunk2 gave warnings:<br>
The web page has long paragraphs with no \n newlines inside them. Remember separator="\n", it only cuts at newlines. So when it finds a paragraph with no newlines that's 1544 characters long, it can't split it and has to keep it as one big chunk despite the 200 limit.<br>
The PDF had lots of newlines (academic paper formatting) so it split cleanly. The webpage didn't.<br>

In [54]:
#By default, the CharacterTextSplitter uses a single newline ("\n") as its separator, but we can change this to any character or string we want.
#For example:
text_splitter2 = CharacterTextSplitter(chunk_size=200, chunk_overlap=20, separator="i")
chunk3 = text_splitter2.split_documents(document)
print(len(chunk3))
chunk3[5].page_content
#Here we used the alphabet 'i' as seperator

145


'stant for \nindividuals seeking guidance and support in these critical areas. \nMind Guide lever ages the capabilities of LangChain  and its \nChatModels, specifically Chat OpenAI, as the bedrock of'

# Trying All Three types of Text Splitters:

In [39]:
from langchain.text_splitter import CharacterTextSplitter, RecursiveCharacterTextSplitter, TokenTextSplitter

sample_text = [Document(page_content="""LangChain is a framework for building LLM applications.
It provides tools for prompt templates, chains, and agents.
You can load PDFs, websites, and other documents.
Then split them into chunks for RAG applications.
RAG stands for Retrieval Augmented Generation.""")]

# Splitter 1 - CharacterTextSplitter
# Splits ONLY at the separator character (\n here)
# If no \n found, keeps chunk as-is even if too big
char_splitter = CharacterTextSplitter(
    chunk_size=100,
    chunk_overlap=10,
    separator="\n"
)
char_chunks = char_splitter.split_documents(sample_text)
print("=== CharacterTextSplitter ===")
print(f"Number of chunks: {len(char_chunks)}")
for i, chunk in enumerate(char_chunks):
    print(f"Chunk {i}: '{chunk.page_content}'")
    print(f"Length: {len(chunk.page_content)}\n")

=== CharacterTextSplitter ===
Number of chunks: 4
Chunk 0: 'LangChain is a framework for building LLM applications.'
Length: 55

Chunk 1: 'It provides tools for prompt templates, chains, and agents.'
Length: 59

Chunk 2: 'You can load PDFs, websites, and other documents.
Then split them into chunks for RAG applications.'
Length: 99

Chunk 3: 'RAG stands for Retrieval Augmented Generation.'
Length: 46



In [40]:
# Splitter 2 - RecursiveCharacterTextSplitter
# Tries multiple separators in order: \n\n, \n, " ", ""
# Much smarter - keeps trying smaller splits until size is met
recursive_splitter = RecursiveCharacterTextSplitter(
    chunk_size=100,
    chunk_overlap=10
)
recursive_chunks = recursive_splitter.split_documents(sample_text)
print("=== RecursiveCharacterTextSplitter ===")
print(f"Number of chunks: {len(recursive_chunks)}")
for i, chunk in enumerate(recursive_chunks):
    print(f"Chunk {i}: '{chunk.page_content}'")
    print(f"Length: {len(chunk.page_content)}\n")

=== RecursiveCharacterTextSplitter ===
Number of chunks: 4
Chunk 0: 'LangChain is a framework for building LLM applications.'
Length: 55

Chunk 1: 'It provides tools for prompt templates, chains, and agents.'
Length: 59

Chunk 2: 'You can load PDFs, websites, and other documents.
Then split them into chunks for RAG applications.'
Length: 99

Chunk 3: 'RAG stands for Retrieval Augmented Generation.'
Length: 46



In [51]:
%%capture
!pip install tiktoken --user

In [52]:
# Splitter 3 - TokenTextSplitter
# Splits by tokens (how AI models actually count text)
# 1 token ≈ 4 characters roughly
from langchain.text_splitter import TokenTextSplitter
token_splitter = TokenTextSplitter(
    chunk_size=20,  # 20 tokens ≈ ~80 characters
    chunk_overlap=5
)
token_chunks = token_splitter.split_documents(sample_text)
print("=== TokenTextSplitter ===")
print(f"Number of chunks: {len(token_chunks)}")
for i, chunk in enumerate(token_chunks):
    print(f"Chunk {i}: '{chunk.page_content}'")
    print(f"Length: {len(chunk.page_content)}\n")

=== TokenTextSplitter ===
Number of chunks: 4
Chunk 0: 'LangChain is a framework for building LLM applications.
It provides tools for prompt templates,'
Length: 95

Chunk 1: ' tools for prompt templates, chains, and agents.
You can load PDFs, websites, and'
Length: 81

Chunk 2: 's, websites, and other documents.
Then split them into chunks for RAG applications.
'
Length: 84

Chunk 3: ' RAG applications.
RAG stands for Retrieval Augmented Generation.'
Length: 65



## Concept for Text Splitters:

If we imagine this is our text:<br>
ABCDEFGHIJKLMNOPQRSTUVWXYZ<br>
With chunk_size=10 and chunk_overlap=3:<br>
Chunk 1: ABCDEFGHIJ<br>
Chunk 2: HIJKLMNOPQ      ← starts 3 back (HI J is overlap)<br>
Chunk 3: OPQRSTUVWX      ← starts 3 back (OPQ is overlap)<br>
Chunk 4: VWXYZ<br>
<br>
Now with real text:<br>
Our text:<br>
LangChain is a framework.<br>
It provides tools.<br>
You can load PDFs.<br>
RAG stands for retrieval.<br>
With chunk_size=50 and separator="\n":<br>
The splitter reads line by line and asks:<br>
<br>
"Can I add this next line without exceeding 50 characters?"<br>
<br>
"LangChain is a framework." → 25 chars, so we can add it<br>
"It provides tools." → 25+19 = 44 chars, so we can add it<br>
"You can load PDFs." → 44+18 = 62 chars, which is too big than '50'!<br>
So Chunk 1 is sealed:<br>
Chunk 1: "LangChain is a framework.<br>
It provides tools."<br>
Then it backs up by chunk_overlap characters and starts Chunk 2 from there.<br>
......<br>
So Chunks:<br>
Chunk 1: LangChain is a framework. It provides<br>
Chunk 2: provides tools. You can load<br>
Chunk 3: load PDFs. RAG stands<br>

## Last Practice:


In [55]:
from langchain_core.documents import Document
from langchain_community.document_loaders import PyPDFLoader, WebBaseLoader
from langchain.text_splitter import CharacterTextSplitter, RecursiveCharacterTextSplitter

# To load the LangChain paper
paper_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf"
pdf_loader = PyPDFLoader(paper_url)
pdf_document = pdf_loader.load()

# Load content from LangChain website
web_url = "https://python.langchain.com/v0.2/docs/introduction/"
web_loader = WebBaseLoader(web_url)
web_document = web_loader.load()

# Create two different text splitters
splitter_1 = CharacterTextSplitter(chunk_size=300, chunk_overlap=30, separator="\n")
splitter_2 = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=50, separators=["\n\n", "\n", ". ", " ", ""])
#This is something new, we'll talk about this down

# Apply both splitters to the PDF document
chunks_1 = splitter_1.split_documents(pdf_document)
chunks_2 = splitter_2.split_documents(pdf_document)

# Define a function to display document statistics
def display_document_stats(docs, name):
    """Display statistics about a list of document chunks"""
    total_chunks = len(docs)
    total_chars = sum(len(doc.page_content) for doc in docs)
    avg_chunk_size = total_chars / total_chunks if total_chunks > 0 else 0
    
    # Count unique metadata keys across all documents
    all_metadata_keys = set()
    for doc in docs:
        all_metadata_keys.update(doc.metadata.keys())
    
    # Print the statistics
    print(f"\n=== {name} Statistics ===")
    print(f"Total number of chunks: {total_chunks}")
    print(f"Average chunk size: {avg_chunk_size:.2f} characters")
    print(f"Metadata keys preserved: {', '.join(all_metadata_keys)}")
    
    if docs:
        print("\nExample chunk:")
        example_doc = docs[min(5, total_chunks-1)]  # Get the 5th chunk or the last one if fewer
        print(f"Content (first 150 chars): {example_doc.page_content[:150]}...")
        print(f"Metadata: {example_doc.metadata}")
        
        # Calculate length distribution
        lengths = [len(doc.page_content) for doc in docs]
        min_len = min(lengths)
        max_len = max(lengths)
        print(f"Min chunk size: {min_len} characters")
        print(f"Max chunk size: {max_len} characters")

# Display stats for both chunk sets
display_document_stats(chunks_1, "Splitter 1")
display_document_stats(chunks_2, "Splitter 2")


=== Splitter 1 Statistics ===
Total number of chunks: 95
Average chunk size: 266.07 characters
Metadata keys preserved: page, source

Example chunk:
Content (first 150 chars): comprehensive support within the field of mental health. 
Additionally, the paper discusses the implementation of 
Streamlit to enhance the user ex pe...
Metadata: {'source': 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf', 'page': 0}
Min chunk size: 65 characters
Max chunk size: 299 characters

=== Splitter 2 Statistics ===
Total number of chunks: 57
Average chunk size: 452.93 characters
Metadata keys preserved: page, source

Example chunk:
Content (first 150 chars): with severe intellectual disorders do no longer have get entry 
to the necessary remedy they require. This remedy gap 
intensifies the weight of intel...
Metadata: {'source': 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf'

# Something to learn:
Check out the line: <br>
splitter_2 = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=50, separators=["\n\n", "\n", ". ", " ", ""])<br>
...This was something new..<br>
This is a priority list of where to cut. The splitter tries them in order:<br>
First try:  "\n\n"  (double newline = paragraph break)<br>
If chunk still too big...<br>
Second try: "\n"    (single newline = line break)<br>
If chunk still too big...<br>
Third try:  ". "    (period + space = sentence end)<br>
If chunk still too big...<br>
Fourth try: " "     (single space = word boundary)<br>
If chunk still too big...<br>
Last resort: ""     (cut anywhere, even mid-word)<br>
...<br>
Why this order makes sense:<br>
It tries to keep the most meaningful units together first:<br>
Paragraphs together (best)<br>
Lines together (good)<br>
Sentences together (okay)<br>
Words together (last resort)<br>
Cut anywhere (desperate)<br>
...<br>
Its something like:<br>
First try cutting between paragraphs<br>
If still too big, cut between sentences<br>
If still too big, cut between words<br>
If still too big, cut mid-word<br>
Always trying to preserve meaning as much as possible.<br>
..<br>
That's why RecursiveCharacterTextSplitter is smarter than CharacterTextSplitter — it has multiple fallback strategies instead of just one.<br>